text/0000-async-stream.md

```diff
@@ -0,0 +1,883 @@
+elements, and once they've all been exhausted, will yield `None` to indicate
+that iteration is finished. If we're waiting on something asynchronous to
+resolve, the future will wait until the stream is ready to yield again.
+
+Individual streams may choose to resume iteration, and so calling [`next`]
+again may or may not eventually yield `Some(Item)` again at some point.
+
+This is similar to the `Future` trait. The `Future::poll_next` method is rarely called 
+directly, it is almost always used to implement other Futures. Interacting
+with futures is done through `async/await`.
+
+A `Stream` by itself is not useful - we need some way to interact with it.
+To interact with it, we need the `next` method - all other interactions
+with `Stream` can be expressed through it. Without the `next` method,
+streams cannot be consumed.
+
+### Usage
+
+Continuing the example of `Stream` implemented on a struct called `Counter`, the user would interact with the stream like so:
+
+```rust
+let mut counter = Counter::new();
+
+let x = counter.next().await.unwrap();
+println!("{}", x);
+
+let x = counter.next().await.unwrap();
+println!("{}", x);
+
+let x = counter.next().await.unwrap();
+println!("{}", x);
+
+let x = counter.next().await.unwrap();
+println!("{}", x);
+
+let x = counter.next().await.unwrap();
+println!("{}", x);
+#
+}
+```
+
+This would print `1` through `5`, each on their own line.
+
+### Why does next require Self:Unpin?
+
+When drafting this RFC, there was a [good deal of discussion](https://github.com/rust-lang/wg-async-foundations/pull/15#discussion_r452482084) around why the `next` method requires `Self:Unpin`.
+
+In particular, there was concern around how this would affect the use of generators in the future. Generators are discussed in the [Future possiblilities](future-possibilities) section later in this RFC.
+
+To understand this, it helps to take a closer look at the definition of `Next` (this struct is further discussed later in this RFC) in the [futures-util crate](https://docs.rs/futures-util/0.3.5/src/futures_util/stream/stream/next.rs.html#10-12).
+
+```rust
+pub struct Next<'a, St: ?Sized> {
+    stream: &'a mut St,
+}
+```
+Since `Stream::poll_next` takes a pinned reference, the next future needs `S` to be `Unpin` in order to safely construct a `Pin<&mut S>` from a `&mut S`.
+
+An alternative approach we could take would be to have the `next` method take `Pin<&mut S>`, rather than `&mut S`. However, this would require pinning even when the type is `Unpin`. The current approach requires pinning only when the type is not `Unpin`.
+
+We currently do see some `!Unpin` streams in practice (including in the [futures-intrusive crate](https://github.com/Matthias247/futures-intrusive/blob/master/src/channel/mpmc.rs#L565-L625)). We also see `stream.then(|_| async {})` resulting in an `!Unpin` stream. Where `!Unpin` streams will become important is when we introduce async generators, as discussed in [future-possibilities].
```

Comment:
Given that we expect to see _more_ `!Unpin` streams going forward, I'd like to see the RFC expand on how `next` interacts with things that are `Unpin` vs `!Unpin`. Some brief code examples in this section would be very helpful I think.

Reply:
A sample would be good. Something like

```rust
let stream = non_unpin_stream();
pin_mut!(stream);  // or use Box::pin
while let Some(item) = stream.next().await {
  ...
}
```

should convey the idea.

Reply:
Added!